kong-builders/12-01-2021/Makefile

```diff
@@ -0,0 +1,101 @@
+
+GCP_PROJECT_ID ?= $(shell gcloud config list --format 'value(core.project)')
+GKE_BASE_MACHINE_TYPE ?= e2-standard-2
+
+CLUSTER_NAME=kong-builders
+CLUSTER_ZONE=us-east1-c
+
+CURRENT_WORKING_DIR=$(shell pwd)
+WHO_AM_I=$(shell whoami)
+TIMESTAMP=$(shell date)
+
+DEMO_NAMESPACE ?= kong-builders
+
+pad=$(printf '%0.1s' "-"{1..80})
+
+ISTIO_VERSION ?= 1.12.0
+
+.DEFAULT_GOAL := help
+
+istio-install: gke-check-dependencies
+	@echo "${GREEN}Installing Istio with default profile⛵${RESET}"
+	@echo "To list all profiles run ${BOLD}istioctl profile list${RESET}"
+
+	if [ -d "./istio-${ISTIO_VERSION}" ]; then \
+		echo "${REST}${BLUE}Found ./istio-${ISTIO_VERSION} ${REST}"; \
+	else \
+		echo "No Istio installation found"; \
+		echo "${GREEN}Downloading Istio...${RESET}" ;\
+		@curl -s -L https://istio.io/downloadIstio | ISTIO_VERSION=${ISTIO_VERSION} sh - ;\
+	fi
+	install -m 755 ./istio-${ISTIO_VERSION}/bin/istioctl ~/bin/istioctl; 
+	istioctl x precheck
+	istioctl install --set profile=default -y
+
+	@echo ""
+
+istio-remove: check-dependencies
+	kubectl delete namespaces istio-system
+
+demo-app-deploy: check-dependencies
+	@echo "${GREEN}Deploying Quote Services (Back to Future and chuck Norris) to ${BLUE}${DEMO_NAMESPACE}${RESET}"
+
+	kubectl create namespace ${DEMO_NAMESPACE}     
+	kubectl label namespace ${DEMO_NAMESPACE} istio-injection=enabled
+	kubectl -n ${DEMO_NAMESPACE} apply -f 2_backtofuture_quote_service.yaml
+	kubectl -n ${DEMO_NAMESPACE} wait --for=condition=Available deployment back2future-quote-service
+	kubectl -n ${DEMO_NAMESPACE} apply -f 2_chuck_quote_service.yaml
+	kubectl -n ${DEMO_NAMESPACE} wait --for=condition=Available deployment chuck-quote-service
+
+	@echo ""
+	
+demo-app-gateway: check-dependencies
+	@echo "${GREEN}Installing Gateway resource to ${BLUE}${DEMO_NAMESPACE}${RESET}"
+	kubectl -n ${DEMO_NAMESPACE} apply -f 1_gateway.yaml
+	@echo ""
+
+demo-app-virtual-services: check-dependencies
+	@echo "${GREEN}Installing Virtual Services resources to ${BLUE}${DEMO_NAMESPACE}${RESET}"
+	kubectl -n ${DEMO_NAMESPACE} apply -f 3_backtofuture_virtual_service.yaml
+	kubectl -n ${DEMO_NAMESPACE} apply -f 3_chuck_virtual_service.yaml
+	@echo ""
+
+demo-app-remove: check-dependencies
+	kubectl delete namespaces ${DEMO_NAMESPACE}
+	
+autopilot: check-dependencies istio-install demo-app-deploy demo-app-gateway demo-app-virtual-services wait-for-ip
+
+cleanup: demo-app-remove istio-remove
+
+wait-for-ip:
+	until kubectl get svc --namespace istio-system istio-ingressgateway --output=jsonpath='{.status.loadBalancer.ingress[0].ip}' > /dev/null; do echo "waiting 10 sec for external IP of LoadBalancer..."; sleep 10; done;
+
+test: IP = $(shell kubectl get svc --namespace istio-system istio-ingressgateway -ojson | jq -r '.status.loadBalancer.ingress[0].ip')
+
+test: 
+	@echo "🛫 Let's fly..."
+	@$(call check-dependency,http)
+	http -v ${IP}
+	http ${IP} Host:chuck.dev
+	http ${IP} Host:backtofuture.dev
```

Comment:
FWIW you might want to consider adding tests for these demos in a language like Python or Go:

- https://docs.python.org/3/library/unittest.html#basic-example
- https://pkg.go.dev/testing

This will give you structured access to the Kubernetes API as well as more robust HTTP libraries for testing these responses.

Not sure how important this is to you, it was just a thought I had that so that you can get better feedback about changes you make and at least in the above two languages, the setup is really simple/basic.

Reply:
Poorly named Makefile target. It is not actually test, just something that I would show in my demo. Essentially `autopilot` target is a script for my demo during a presentation. 
Thank you. I will check those links as well.